parlai/nn/lr_scheduler.py

```diff
@@ -0,0 +1,217 @@
+        )
+    elif opt.get('lr_scheduler') == 'invsqrt':
+        scheduler = InvSqrtLRScheduler(
+            opt, optimizer, states, hard_reset, patience, decay
+        )
+    elif opt.get('lr_scheduler') == 'cosine':
+        scheduler = CosineLRScheduler(
+            opt, optimizer, states, hard_reset, patience, decay
+        )
+    elif opt.get('lr_scheduler') == 'linear':
+        scheduler = LinearLRScheduler(
+            opt, optimizer, states, hard_reset, patience, decay
+        )
+    else:
+        raise ValueError(
+            "Don't know what to do with lr_scheduler '{}'".format(
+                opt.get('lr_scheduler')
+            )
+        )
+
+    # time to load LR state from the checkpoint, if possible.
+    if (
+        # there is already an old LR scheduler saved on disk
+        states
+        and
+        # and the old LR scheduler is different
+        states.get('lr_scheduler_type') != opt['lr_scheduler']
+        and
+        # and we're not already using a fresh scheduler
+        not hard_reset
+    ):
+        # the LR scheduler changed, start things fresh
+        warn_once("LR scheduler is different from saved. Starting fresh!")
+        hard_reset = True
+
+    if hard_reset:
+        # We're not going to use the LR schedule, let's just exit
+        return
+    else:
+        scheduler.load_state(states)
+        # do the actual loading (if possible)
+
+
+class ParlAILRScheduler(object):
+    def __init__(self, *args, **kwargs):
+        if self.fp16:
+            # lr schedulers don't work with apex, they expect the "real" optimizer
+            optimizer = optimizer.optimizer
+
+        warmup_updates = opt.get('warmup_updates', -1)
+        updates_so_far = states.get('number_training_updates', 0)
+        if warmup_updates > 0 and (updates_so_far < warmup_updates or hard_reset):
+
+            def _warmup_lr(step):
+                start = opt['warmup_rate']
+                end = 1.0
+                progress = min(1.0, step / opt['warmup_updates'])
+                lr_mult = start + (end - start) * progress
+                return lr_mult
+
+            self.warmup_scheduler = optim.lr_scheduler.LambdaLR(optimizer, _warmup_lr)
+        else:
+            self.warmup_scheduler = None
+
+    def load_state(self, states):
+        if 'number_training_updates' in states:
+            self._number_training_updates = states['number_training_updates']
+        if self.scheduler and 'lr_scheduler' in states:
+            self.scheduler.load_state_dict(states['lr_scheduler'])
+        if states.get('warmup_scheduler') and getattr(self, 'warmup_scheduler', None):
+            self.warmup_scheduler.load_state_dict(states['warmup_scheduler'])
+
+    def _is_lr_warming_up(self):
+        """Check if we're warming up the learning rate."""
+        return (
+            self.warmup_scheduler is not None
+            and self._number_training_updates <= self.opt['warmup_updates']
+        )
+
+    def step():
```

Comment:
I wouldn't mind both of these marked with `@abc.abstractmethod`, which would require a small extra bit of code (pass) but make it more explicit and catch some errors.

Reply:
also this function should take an explicit step number (and self)